src/cripts/README.md

```diff
@@ -0,0 +1,453 @@
+  - [4.2. cachekey](#42-cachekey)
+  - [4.3. header\_rewrite](#43-header_rewrite)
+  - [4.4. regex\_remap](#44-regex_remap)
+  - [4.5. geoip\_acl and maxmind\_acl](#45-geoip_acl-and-maxmind_acl)
+  - [4.6. tcpinfo](#46-tcpinfo)
+
+# 2. Building Cripts
+
+Cripts needs the `{fmt}` and `PCRE2` libraries and include files. We currently only
+build cripts when explicitly enabled, and only using `cmake``.
+
+## 2.1. Building a Cript
+
+At the moment, building Cripts is the same as building any other ATS plugin, using
+either existing build system or tooling such as `tsxs`.
+
+```
+tsxs -lcript -o cript_test.so cript_test.cc
+```
+
+This will read the Cript file cript_test.cc and produce a cript_test.so plugin, that can be
+loaded in ATS via remap.config:
+
+```
+map https://example.com https://origin.example.com @plugin=cript_test.so
+```
+
+## 2.2. Lint: Validating a Cript
+
+TBD: I have the beginning of this tool, will land it later.
+
+## 2.3. Clang-tidy
+
+A custom clang-tidy configuration is provided with this patch, and I've run all
+code through clang-tidy with this configuration.
+
+# 3. Writing Cripts
+
+Cripts follow the same basic model of how ATS splits up transaction processing
+into what we call "hooks". A hook is essentially a callback mechanism, where
+custom code can be injected into the ATS core via plugins.
+
+The Cript language itself is essentially C++17, except it imposes some serious,
+but important, limitations on what can and can not be used. Albeit we call this
+a scripting language, it's truly compiled into regular, reloadable ATS plugins.
+
+To start off with, we'll show a very basic Cript, to get an idea of what to
+expect:
+
+```
+// The primary include file, this has to always be included
+#include <Cript/Preamble.hpp>
+
+do_send_response()
+{
+  borrow req  = Client::Request::get();
+  borrow resp = Client::Response::get();
+
+  if (req["X-Miles"] != "") {
+      resp["X-Miles"] = req["X-Miles"]; // Echo back the request header
+  }
+}
+
+do_remap()
+{
+  borrow req = Client::Request::get();
+
+  req["@receipt"] = "PropertyX";
+}
+
+#include <Cript/Epilogue.hpp>
+```
+
+Don't worry about the exact details here, we will discuss them all further
+down this documentation. There are however two critical pieces here to remember:
+
+* All Cript's must have the Preamble and Epilogue include's as above
+* You can have your own C/C++ function definitions if you like, but the predefined
+  callbacks (such as `do_remap()`) have fixed names
+
+## 3.1. Data types
+
+Cript will make a best-effort to hide data types from the users as much as possible.
+As such, it's highly recommended to use the "auto" style when declaring variables.
+Cript being C/C++, it's impossible to hide everything, and to optimize integration
+with ATS, we have a few quirks.
+
+For example, there are two types of strings which can be returned, regular `strings`
+and something called `string_view`. The latter are immutable representations of
```

Comment:
We should should clarify here that we're referring to `Cript::string_view` rather than `std::string_view`. Elsewhere in this paragraph, `string` means `std::string`, so it would be natural for someone to assume `string_view` is simply the `std` version. Might want to say how `Cript::string_view` is different from the `std` as a user will be curious.